freshclam fix for RHEL7

coreone opened the pull request:
Added a sysconfig file for freshclam in RHEL7.  RHEL7 comes by default with an option in `/etc/sysconfig/freshclam` which disables freshclam.  So, in order to make freshclam work, I needed to start managing that file in this module for RHEL7.

coreone commented:
Just an update on this.  What `/etc/sysconfig/freshclam` fixes is actually `freshclam-sleep`, which is what cron runs daily to do updates on RHEL7.  That script checks that _FRESHCLAM_DELAY_ value if it exists before running freshclam with some random _sleep_ to add some splay to the updating.

edestecd commented on manifests/freshclam.pp:
```diff
@@ -33,7 +45,10 @@
+      subscribe  => [
+        File['freshclam.conf'],
+        File['freshclam_sysconfig'],
+      ],
```
This may fail if not on RedHat 7.
Can you instead do something like lines 55-57?

If both freshclam_service and freshclam_sysconfig then notify...

edestecd commented on manifests/init.pp:
```diff
@@ -68,7 +70,7 @@
   $_freshclam_options = merge($clamav::params::freshclam_default_options, $freshclam_options)
-
+  validate_string($freshclam_delay)
 
```
How about a`validate_absolute_path($freshclam_sysconfig)`

edestecd commented on templates/sysconfig/freshclam.erb:
```diff
@@ -0,0 +1,5 @@
+#
+# Managed by Puppet
+#
+
```
Can you include the standard header like the other template?
https://github.com/broadinstitute/puppet-clamav/blob/525c4dc425a1664a29dcf6fb4033186b9d8b0316/templates/clamav.conf.erb

edestecd commented on manifests/freshclam.pp:
```diff
@@ -25,6 +26,17 @@
   }
 
+  if ($::osfamily == 'RedHat') and (versioncmp($::operatingsystemmajrelease, '7') >= 0) {
+    file { 'freshclam_sysconfig':
```
How about changing the condition to just:

``` puppet
if $clamav::freshclam_sysconfig {
}
```

Since you already check the OS and set that in params...

edestecd commented:
Can you also add some checks to the spec test here for your new file?
https://github.com/edestecd/puppet-clamav/blob/master/spec/classes/clamav_spec.rb#L71

This way we can validate it is actually being managed on the correct OS and version...

edestecd commented:
@coreone Thank you for your contribution.  I have had several mention that we need to manage this file to fix the issue you describe.  I think this would be a great contribution to the module.
I just had a few suggestions above.  Please let me know if you need any help with my suggestions.

coreone commented:
@edestecd I made the changes you suggested.  The one that was slightly different was the subscribe piece.  Take a look and let me know if there are any further issues.

edestecd commented on spec/classes/clamav_spec.rb:
```diff
@@ -72,6 +72,7 @@
+              it { is_expected.to contain_file('freshclam_sysconfig') }
             end
             it { is_expected.to contain_file('freshclam.conf') }
             it { is_expected.not_to contain_service('freshclam') }
```
Can you add checks for not containing freshclam_sysconfig in unsupported oses?

``` ruby
it { is_expected.not_to contain_file('freshclam_sysconfig') }
```

edestecd commented on manifests/params.pp:
```diff
@@ -53,6 +53,8 @@
         'DatabaseMirror' => 'database.clamav.net',
       }
+      $freshclam_sysconfig = '/etc/sysconfig/freshclam'
+      $freshclam_delay     = 'disabled-warn'
```
`disabled-warn` is what the rpm puts in here by default right?
This effectively disables updates...  What do you change this to to get it working?
Should we maybe put something better here, so that it works out of the box.

I think the general consensus is that the defaults for RHEL 7 are pretty bad and should be changed to something that produces a working system out of the box...
Were working on the config file defaults as well to this end.

edestecd commented:
The subscribe piece is fine.  The end result is the same.
Sorry for the delay.  As soon as you do the last two things, I will merge.  Everything else looks great.

Thanks for your work!  Several people have asked for this fix and I am looking forward to having it in the next release.

edestecd commented:
@coreone The subscribe piece is fine. The end result is the same.
Sorry for the delay. As soon as you do the last two things, I will merge. Everything else looks great.

Thanks for your work! Several people have asked for this fix and I am looking forward to having it in the next release.

coreone commented:
All set.  Added the tests for everything but RHEL7, which is the only place that sysconfig file is used.  Also changed the default for `freshclam_delay` to _undef_, since that will cause the script to choose a random splay based on `hostid`, which should make it work "out of the box".  Either empty string or the string **random** in that variable in the sysconfig file will cause this behavior.  Entering a number for this field will base the splay off that number, so empty string seemed the easiest.

coreone commented:
@edestecd if all looks good, do you want me to do a squash on this branch to clean up the history to one all encompassing commit?

edestecd commented:
@coreone I was just thinking of asking you to squash... lol
That would be great.

coreone commented:
@edestecd all squashed
